### PR TITLE
Netplay: Properly save hosting settings

### PR DESCRIPTION
**Pull request by Zexaron**
Addresses issue: [11398](https://bugs.dolphin-emu.org/issues/11398)

EDIT: Apparently some these settings need to be synced with other clients, and it will probably need some actual netplay testing.

EDIT2: The bug report mentions "friends" ... note sure about that yet, will take a look.

### CODE REVIEW

File: Source/Core/Core/Config/NetplaySettings.cpp [hunk 1]
```diff
@@ -35,4 +35,22 @@ const ConfigInfo<bool> NETPLAY_USE_UPNP{{System::Main, "NetPlay", "UseUPNP"}, fa
 
 const ConfigInfo<bool> NETPLAY_ENABLE_QOS{{System::Main, "NetPlay", "EnableQoS"}, true};
 
+// Host.NetPlay
```
Comment (Techjar) [inline Source/Core/Core/Config/NetplaySettings.cpp]: This isn't the semantics of these comments, we're still in `Main.NetPlay`. Having the host prefix on all of these options is probably redundant anyways, as it's like there's a non-host variant of them.

File: Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp [hunk 2]
```diff
@@ -71,6 +71,25 @@ NetPlayDialog::NetPlayDialog(QWidget* parent)
...
+
+  m_buffer_size_box->setValue(buffer_size);
+  m_save_sd_box->setChecked(write_save_sdcard_data);
+  m_load_wii_box->setChecked(load_wii_save);
+  m_sync_save_data_box->setChecked(sync_saves);
+  m_record_input_box->setChecked(record_inputs);
+  m_reduce_polling_rate_box->setChecked(reduce_polling_rate);
+  m_strict_settings_sync_box->setChecked(strict_settings_sync);
+  m_host_input_authority_box->setChecked(host_input_authority);
+
```
Comment (Techjar) [inline Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp]: This should sort out the option state syncing with the client/server instances.

```c++
  auto server = Settings::Instance().GetNetPlayServer();
  if (server)
    server->SetHostInputAuthority(host_input_authority);
  if (host_input_authority)
  {
    auto client = Settings::Instance().GetNetPlayClient();
    if (client)
      client->AdjustPadBufferSize(buffer_size);
  }
  else
  {
    if (server)
      server->AdjustPadBufferSize(buffer_size);
  }
```

File: Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp [hunk 3]
```diff
@@ -71,6 +71,25 @@ NetPlayDialog::NetPlayDialog(QWidget* parent)
   CreateChatLayout();
   CreatePlayersLayout();
   CreateMainLayout();
+
+  int buffer_size = Config::Get(Config::NETPLAY_HOST_BUFFER_SIZE);
```
Comment (Techjar) [inline Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp]: All of this needs to be moved to `CreateMainLayout`, otherwise it will just segfault immediately.

Comment (Techjar) [inline Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp]: These can be const.

File: Source/Core/Core/Config/NetplaySettings.cpp (hunk 1 above)
Comment (Zexaron) [reply Source/Core/Core/Config/NetplaySettings.cpp]: Indeed I thought about that half-way through it, I don't have strong opinion but it's a bit more clear, do you care more about the prefixes in the code or in the .ini ?

Comment (Techjar) [reply Source/Core/Core/Config/NetplaySettings.cpp]: Both honestly. I don't really see the need to clarify that these are host-only options, as it's kind of already implied by the fact that the checkboxes aren't visible to non-hosts. Also, buffer size and record inputs aren't host-only options.

File: Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp [hunk 4]
```diff
@@ -910,6 +941,22 @@ std::shared_ptr<const UICommon::GameFile> NetPlayDialog::FindGameFile(const std:
   return nullptr;
 }
 
+void NetPlayDialog::SaveSettings()
+{
+  Config::SetBaseOrCurrent(Config::NETPLAY_HOST_BUFFER_SIZE,
```
Comment (Techjar) [inline Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp]: Use `SetBase` instead, as we want these to persist outside of any currently active game session.

**Comment by Techjar** [comment]
The following diffs should make the option states sync correctly on NetPlay start. Add a config called `NETPLAY_CLIENT_BUFFER_SIZE` as well for this. I also assume that the HOST_ prefixes have been removed, though change as necessary.

```diff
+++ Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp
@@ -943,8 +943,16 @@
 
 void NetPlayDialog::SaveSettings()
 {
-  Config::SetBaseOrCurrent(Config::NETPLAY_HOST_BUFFER_SIZE,
-                           static_cast<u8>(m_buffer_size_box->value()));
+  if (m_host_input_authority)
+  {
+    Config::SetBase(Config::NETPLAY_CLIENT_BUFFER_SIZE,
+                    static_cast<u8>(m_buffer_size_box->value()));
+  }
+  else
+  {
+    Config::SetBase(Config::NETPLAY_BUFFER_SIZE,
+                    static_cast<u8>(m_buffer_size_box->value()));
+  }
   Config::SetBaseOrCurrent(Config::NETPLAY_HOST_WRITE_SAVE_SDCARD_DATA, m_save_sd_box->isChecked());
   Config::SetBaseOrCurrent(Config::NETPLAY_HOST_LOAD_WII_SAVE, m_load_wii_box->isChecked());
   Config::SetBaseOrCurrent(Config::NETPLAY_HOST_SYNC_SAVES, m_sync_save_data_box->isChecked());
```
```diff
+++ Source/Core/DolphinQt/MainWindow.cpp
@@ -1092,6 +1092,8 @@
                           QObject::tr("A NetPlay Session is already in progress!"));
     return false;
   }
+  
+  auto server = Settings::Instance().GetNetPlayServer();
 
   // Settings
   const std::string traversal_choice = Config::Get(Config::NETPLAY_TRAVERSAL_CHOICE);
@@ -1099,10 +1101,10 @@
 
   std::string host_ip;
   u16 host_port;
-  if (Settings::Instance().GetNetPlayServer())
+  if (server)
   {
     host_ip = "127.0.0.1";
-    host_port = Settings::Instance().GetNetPlayServer()->GetPort();
+    host_port = server->GetPort();
   }
   else
   {
@@ -1114,9 +1116,17 @@
   const std::string traversal_host = Config::Get(Config::NETPLAY_TRAVERSAL_SERVER);
   const u16 traversal_port = Config::Get(Config::NETPLAY_TRAVERSAL_PORT);
   const std::string nickname = Config::Get(Config::NETPLAY_NICKNAME);
+  const bool host_input_authority = Config::Get(Config::NETPLAY_HOST_INPUT_AUTHORITY);
+  
+  if (server)
+  {
+    server->SetHostInputAuthority(host_input_authority);
+    if (!host_input_authority)
+      server->AdjustPadBufferSize(Config::Get(Config::NETPLAY_BUFFER_SIZE));
+  }
 
   // Create Client
-  const bool is_hosting_netplay = Settings::Instance().GetNetPlayServer() != nullptr;
+  const bool is_hosting_netplay = server != nullptr;
   Settings::Instance().ResetNetPlayClient(new NetPlay::NetPlayClient(
       host_ip, host_port, m_netplay_dialog, nickname,
       NetPlay::NetTraversalConfig{is_hosting_netplay ? false : is_traversal, traversal_host,
@@ -1127,9 +1137,12 @@
     NetPlayQuit();
     return false;
   }
+  
+  if (host_input_authority)
+    Settings::Instance().GetNetPlayClient()->AdjustPadBufferSize(Config::Get(Config::NETPLAY_CLIENT_BUFFER_SIZE));
 
-  if (Settings::Instance().GetNetPlayServer() != nullptr)
-    Settings::Instance().GetNetPlayServer()->SetNetPlayUI(m_netplay_dialog);
+  if (server != nullptr)
+    server->SetNetPlayUI(m_netplay_dialog);
 
   m_netplay_setup_dialog->close();
   m_netplay_dialog->show(nickname, is_traversal);
```

**Comment by Zexaron** [comment]
The second diff: 
techjar2.diff:7: trailing whitespace.

techjar2.diff:30: trailing whitespace.

techjar2.diff:48: trailing whitespace.

error: corrupt patch at line 58

**Comment by Zexaron** [comment]
The diff from gist: 
error: patch failed: Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp:958
error: Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp: patch does not apply

**Comment by Zexaron** [comment]
There, should be it, you forgot about adding CLIENT version of Buffer Size into the configs apparently so I did that as well.

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by Techjar** [comment]
Using u8 for the config was causing typecasting issues, since the value is a u32. Makes sense to just change the config to u32, so I did that. I also fixed some odd behavioral issues. Here's the diff to apply: https://gist.githubusercontent.com/Techjar/5dd4dd4516263027f2726d913a20cc19/raw/dba909e7be2d63f0e75be759ed1339b6b53983dc/gistfile1.diff

Once that's done I'm happy to call this ready to merge.

**Comment by Zexaron** [comment]
Diff applied.

**Comment by Techjar** [comment]
I believe you read an old revision of my comment, as there's some stuff missing.

**Comment by Techjar** [comment]
Here's a diff with the thing you missed: https://gist.githubusercontent.com/Techjar/f6dc80c7b4b8895649148a71aee18a54/raw/a1b60e8004b4e55885be45426765efdadaa3babe/gist.diff

**Comment by Techjar** [comment]
Here's a diff with the parts you missed: https://gist.githubusercontent.com/Techjar/c6c52d56669b74c5b1946932124c5424/raw/d1285cf3ea15ec67a1288f2f5429cde3677a46f1/correction.diff

**Comment by Zexaron** [comment]
Diff applied.

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild

**Pull request merged**